File: getDirectoryTree.py
import json
import subprocess
import os

def parse_ls_tree_line(gitTreeLine):
    """Read one line of `git ls-tree` output and produce filename + metadata fields"""
    metadata, filename = gitTreeLine.split('\t')
    mode, obj_type, obj_hash, size = metadata.split()
    return [filename, mode, obj_type, obj_hash, size]

def read_ls_tree_line(tree, dataset, dataset_directory, gitTreeLine, path=''):
    """Read one line of `git ls-tree` and append to the correct buckets of files, symlinks, and objects."""
    filename, mode, obj_type, obj_hash, size = parse_ls_tree_line(
        gitTreeLine)
    # Skip git / datalad files
    if filename.startswith('.git'):
        return
    if filename.startswith('.datalad'):
        return
    if filename == '.gitattributes':
        return

    if obj_type == "blob":
        tree.append({"text": filename, "href": f"https://raw.githubusercontent.com/OpenNeuroDatasets/{dataset}/master/{path}{filename}"})
    if obj_type == "tree":
        tree.append({"text": filename, "state": {"expanded": False}, "nodes": get_repo_files(dataset, dataset_directory, obj_hash, path=f"{path}{filename}/")})



def get_repo_files(dataset, dataset_directory, branch='HEAD', path=''):
    tree = []
    """Read all files in a repo at a given branch, tag, or commit hash."""
    gitProcess = subprocess.Popen(
        ['git', 'ls-tree', '-l', branch], cwd=dataset_directory, stdout=subprocess.PIPE, encoding='utf-8')
    for line in gitProcess.stdout:
        gitTreeLine = line.rstrip()
        read_ls_tree_line(tree, dataset, dataset_directory, gitTreeLine, path)
    return tree

nemar_path = '/Users/dtyoung/Documents/NEMAR/openneuro_datasets'
with open('openneuro_datasets_cont','r') as f:
    for line in f.readlines():
        if not line.__contains__("processed"):
            datasetId = line.strip()
            if not os.path.isfile(f'data/{datasetId}TreeWithLinks.txt'):
                print(f'processing {datasetId}...')
                dataset_directory = f'{nemar_path}/{datasetId}'
                try:
                    gitProcess = subprocess.run(
                        ['datalad', 'install', f'https://github.com/OpenNeuroDatasets/{datasetId}.git'], check=True, cwd=nemar_path, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
                    with open(f'data/{datasetId}TreeWithLinks.txt', 'w') as out:
                        out.write(json.dumps(get_repo_files(datasetId, dataset_directory)))
                except subprocess.CalledProcessError:
                    print(f'Error processing {datasetId}. Move on')

File: getDirectoryTree.php
<?php
function parse_ls_tree_line($gitTreeLine) {
    list ($metadata, $filename) = explode("\t",$gitTreeLine);
    list ($mode, $obj_type, $obj_hash, $size) = preg_split('/\s+/', $metadata);
    return array ($filename, $mode, $obj_type, $obj_hash, $size);
}

function read_ls_tree_line($tree, $dataset, $dataset_directory, $gitTreeLine, $path="") {
    list ($filename, $mode, $obj_type, $obj_hash, $size) = parse_ls_tree_line(
        $gitTreeLine);
    # Skip git / datalad files
    if (substr($filename,0,4) === '.git')
        return $tree;
    if (substr($filename,0,8,) === '.datalad')
        return $tree;
    if (strcmp($filename,'.gitattributes') === 0)
        return $tree;

    if (strcmp($obj_type, 'blob') === 0) {
        array_push($tree, array(
            "text" => $filename,
            "href" => "https://raw.githubusercontent.com/OpenNeuroDatasets/$dataset/master/$path$filename"));
    }
    if (strcmp($obj_type, 'tree') === 0) {
        array_push($tree, array(
            "text" => $filename,
            "state" => array("expanded" => false),
            "nodes" => get_repo_files($dataset, $dataset_directory, $obj_hash, "$path$filename/")));
    }
    return $tree;
}
function get_repo_files($dataset, $dataset_directory, $branch='HEAD', $path='')
{
    $tree = array();
    $gitProcess = proc_open("git ls-tree -l $branch", array(1 => array("pipe","w")), $pipes, $dataset_directory, null);
    if (is_resource($gitProcess)) {
        while (($line = fgets($pipes[1])) !== false) {
            $gitTreeLine = rtrim($line);
            $tree = read_ls_tree_line($tree, $dataset, $dataset_directory, $gitTreeLine, $path);
        }
        proc_close($gitProcess);
    }
    return $tree;
}

$nemar_path = "/Users/dtyoung/Documents/NEMAR/openneuro_eeg";
$file = file('datasets.txt');
foreach ($file as $line) {
    if (!stripos($line, "processed")) {
        echo $line;
        $datasetId = trim($line);
        $dataset_directory = "$nemar_path/$datasetId";
        $out = fopen("data/$datasetId" . "TreeWithLinks.txt", "w");
        fwrite($out,json_encode((get_repo_files($datasetId, $dataset_directory))));
        fclose($out);
    }
}
?>
